utils/yoke/src/cart.rs

```diff
@@ -0,0 +1,120 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Yoke, Yokeable};
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// A [`Cart`] is a type that is acceptable as backing storage in a [`Yoke`].
+///
+/// The essential invariant that must be maintained by implementors is that `Self::Inner` references
+/// obtained from this type via `Self::get_inner()` must continue to be valid for the
+/// complete lifetime of this type (i.e. until destructors run) provided that this type
+/// is never accessed via `&mut` references.
+///
+/// For example, `Rc<Vec<u8>>` and `Rc<[u8]>` are valid [`Cart`]s, however `Rc<RefCell<Vec<u8>>>` is
+/// not, because in the latter type it is possible to use interior mutation to change the data.
```

Comment:
So, again to clarify, this sounds very strongly like a "pinned as long as a `&T` is reachable", but I'm not certain that's what you're saying.

Reply:
Pinned as long as &T::Inner is reachable, yes.